### ISSUE TITLE
Get PR builds to finish in under 5 minutes

### ISSUE THREAD

**Issue by rcoh** [rcoh/angle-grinder#51]
It's now taking ~30 minutes per platform. Goal: PR builds finish in under 5 minutes.

Investigate if we're doing something significantly different than other Rust CLI apps. 

- Maybe it's the cache? I don't think we're really because we use `cross` using it but it takes a significant amount of time to save and load
- We don't really need to build with and without `--release` on branches, but it would be good to keep doing it on master
- Do we need to use `cross` for PR builds?
- Is it worth testing OSX for PR builds? (multi platform doesn't really add much time since they run in parallel though)

Things that matter:
- master builds & runs tests on windows, osx and linux + nightly rust on linux
- Tags produce statically linked binaries

**Comment by jamwaffles** [comment]
It looks like most of that time is spent in Docker containers. Cargo's cache doesn't get copied into them AFAIK, so it will spend the time downloading and compiling every dependency for every arch for every build. Is it enough to `cargo build --target <target arch>`, or should/must the tests be run against that arch?

A bandaid I can think of right now is defining a build matrix in the Travis config so each arch is at least built and tested in parallel. I haven't had good luck exposing Cargo caches to Docker containers, but it might worth a shot with a volume mapping.

**Comment by rcoh** [comment]
For pull requests, I'm fine only running the tests directly on linux without using docker

**Comment by akdor1154** [comment]
I am a bit confused over what `cross` is doing in there tbh. Unless I am really misunderstanding how Travis is configured, it looks like each platform is built on its native OS type, so `cross` isn't really achieving anything at present. (So on linux, host os is linux and cross targets x86_64-unknown-linux-musl; on OS X, host OS is OS X and cross targets x86_64-darwin-whatever-the-rest-of-the-tuple-is.)

I'm not sure if this impacts build time however.

Generally when I need quick CI my first response is "set up CircleCI", not sure if you are up for such a change though. :) You'd need to stick with `cross` then though, afaik it only supports Linux host OSes.

**Comment by rcoh** [comment]
> Additionally, when I was testing, I think the cache is acting per-branch. So updates to a PR would be down to normal times, but each new PR is currently quite a slow build. I'm not sure if this is by design in Travis or if it's configurable (or if it's a bug...)

From the travis docs,
> If a branch does not have its own cache, Travis CI fetches the default branch cache.

I'm not sure if that's something I have to setup manually, but it seems like branches should get a cache on first-built

**Comment by rcoh** [comment]
I made some tweaks in #54 that get PR builds down to 10 minutes, a decent improvement

**Comment by rcoh** [comment]
I have no idea what happened, maybe something on Travis? Anyway, builds now finish in 3 minutes!
